### ISSUE TITLE
The token has been blacklisted - refresh token flow

### ISSUE THREAD

**Issue by IsidroMar95** [sidebase/nuxt-auth#239]
### Environment

------------------------------
- Operating System: `Darwin`
- Node Version:     `v18.13.0`
- Nuxt Version:     `3.2.0`
- Nitro Version:    `2.2.1`
- Package Manager:  `yarn@1.22.19`
- Builder:          `vite`
- User Config:      `app`, `modules`, `components`, `auth`, `headlessui`, `runtimeConfig`
- Runtime Modules:  `@nuxtjs/tailwindcss@6.3.1`, `@sidebase/nuxt-auth@0.4.0-alpha.6`, `nuxt-headlessui@1.1.1`
- Build Modules:    `-`
------------------------------

### Describe the bug

I tried to replicate the [Directus recipe](https://sidebase.io/nuxt-auth/recipes/directus) to implement the JWT refresh token with JWT Laravel API.
I changed the time expiration to 1 minute in order to get the refresh token works, but sometimes works sometimes doesn't. 

```ts
// ~/server/api/auth/[...].ts

import CredentialsProvider from 'next-auth/providers/credentials'
import { NuxtAuthHandler } from '#auth'

const config = useRuntimeConfig()

async function refreshAccessToken(refreshToken: {
  accessToken: string
  accessTokenExpires: string
}) {
  try {
    console.warn('trying to post to refresh token')
    const refreshedTokens = await $fetch<{
      data: {
        access_token: string
        expires_in: number
      }
    } | null>(`${config.API_BASE_URL}/auth/refresh`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${refreshToken.accessToken}`,
      },
    })

    if (!refreshedTokens || !refreshedTokens.data) {
      console.warn('No refreshed tokens')
      throw refreshedTokens
    }

    console.warn('Refreshed tokens successfully')
    return {
      ...refreshToken,
      accessToken: refreshedTokens.data.access_token,
      accessTokenExpires: Date.now() + refreshedTokens.data.expires_in * 1000,
    }
  } catch (error: any) {
    console.warn('Error refreshing token', error.response)
    return {
      ...refreshToken,
      error: 'RefreshAccessTokenError',
    }
  }
}

export default NuxtAuthHandler({
  secret: config.NUXT_SECRET,
  pages: {
    signIn: '/login',
  },

  providers: [
    // @ts-expect-error You need to use .default here for it to work during SSR. May be fixed via Vite at some point
    CredentialsProvider.default({
      name: 'Credentials',
      credentials: {},
      async authorize(credentials: any) {
        try {
          const payload = {
            email: credentials.email,
            password: credentials.password,
          }

          const userTokens = await $fetch<{
            data: { access_token: string; expires_in: number }
          } | null>(`${config.API_BASE_URL}/auth/login`, {
            method: 'POST',
            body: payload,
            headers: {
              'Content-Type': 'application/json',
            },
          })

          const userDetails = await $fetch<{
            data: {
              id: string
              email: string
              roles: []
              user_details: {}
            }
          } | null>(`${config.API_BASE_URL}/auth/me`, {
            method: 'GET',
            headers: {
              'Content-Type': 'application/json',
              Authorization: `Bearer ${userTokens?.data?.access_token}`,
            },
          })

          if (
            !userTokens ||
            !userTokens.data ||
            !userDetails ||
            !userDetails.data
          ) {
            throw createError({
              statusCode: 500,
              statusMessage: 'Next auth failed',
            })
          }

          const user = {
            id: userDetails.data.id,
            email: userDetails.data.email,
            details: userDetails.data.user_details,
            roles: userDetails.data.roles,
            accessToken: userTokens.data.access_token,
            accessTokenExpires: Date.now() + userTokens.data.expires_in * 1000,
          }

          return user
        } catch (error) {
          console.warn('Error logging in', error)
          return null
        }
      },
    }),
  ],

  session: {
    strategy: 'jwt',
  },

  callbacks: {
    async jwt({ token, user, account }) {
      if (account && user) {
        console.warn('JWT callback', { token, user, account })
        return {
          ...token,
          ...user,
        }
      }

      // Handle token refresh before it expires of 2 minutes
      if (token.accessTokenExpires && Date.now() > token.accessTokenExpires) {
        console.warn('Token is expired. Getting a new')
        return refreshAccessToken(token)
      }

      return token
    },
    async session({ session, token }) {
      session.user = {
        ...session.user,
        ...token,
      }

      return session
    },
  },
})
```
The login works, but when I refresh the page I got this error
![CleanShot 2023-02-13@2x](https://user-images.githubusercontent.com/23019968/218561620-f7c336eb-33fd-4685-8944-e1851434ec17.png)

### Additional context

The weird thing is the JWT refresh token works either on Postman or Flutter app mobile, even on Nuxt 2 Auth module. Im using Laravel JWT [tymon/jwt-auth](https://packagist.org/packages/tymon/jwt-auth) on API

![CleanShot 2023-02-13@2x](https://user-images.githubusercontent.com/23019968/218562153-eddde3c7-28bf-419e-b430-feb97d68434b.png)

**Comment by KoenCa** [comment]
Maybe I've run into the same issue on a project for work. I also implemented a credentials provider that has to refresh refresh and access tokens. I came across this issue when I was logged in, closed the tab and came back later and opened the application again. During my investigation I noticed that a race condition can occur when the `refreshAccessToken` method is called. It would basically go like this:

1. Tab is reopened on a secured page (default auth middleware)
    -> Session is fetched from the server
    -> Access token has expired so `refreshAccessToken` is called
    -> The refresh token call succeeds
2. This happens at the same time parallel to point 1: somewhere I use the `useSession` composable for some app logic
   -> Session is fetched from the server
   -> Access token has expired so `refreshAccessToken` is called **with old tokens, because other refreshAccessToken 
        call is also running**
   -> **The refresh token call fails**, because the back-end use refresh token rotation so the old refresh token was invalidated in the previous refresh.
   
I hope this explains the issue properly. I have reduced this, by not using the `getSession` method provided by the `useSession` composable.

While investigation I also came across some similar discussions and issue from next-auth:
- https://github.com/nextauthjs/next-auth/discussions/3940
- https://github.com/nextauthjs/next-auth/discussions/1455
- https://github.com/nextauthjs/next-auth/issues/2129

There should be some way built in that detects if the token is already refreshing and then wait until that completes and use the result of that call.

Some quick fixes that could be done are
- Increase access token expiration time in the back-end
- Don't use refresh token rotation in the back-end
    -> If you only pass your access token to the client, then the refresh token can't be hijacked (correct me if I'm wrong).

**Comment by zoey-kaiser** [comment]
I will close this issue as there has not been any progress. Check out some of the solutions @KoenCa provided. If it still persists feel free to reopen this issue!

**Comment by KoenCa** [comment]
They are not really solutions, but just workarounds. The workarounds reduce the security. The best solution would be that the library (next-auth or nuxt-auth) implements a way that these race conditions can't occur. Basically getSession is called twice or more when you use the useSession hook multiple times. When this happens when the tokens need to be refreshed you get a race condition when you have token blacklisting enabled on your API. 

I think the best solution would be that there can only be one getSession call at a time. When one is in progress, then you could wait on that one and then use its result.

**Comment by KoenCa** [comment]
Race conditions in my situation mainly occured during server side rendering. Since Nuxt 3.4.1 this looks like it was solved, because now the refreshed server session is properly passed to the client after SSR. In the back-end we also implemented a grace period for blacklisted refresh tokens since there can always be some race conditions in browser based apps. A blacklisted token is still valid for x amount of seconds after it has been blacklisted to combat race conditions or concurrency issues.
